hawk/src/main/java/com/orhanobut/hawk/DataHelper.java

```diff
@@ -40,57 +37,74 @@ static DataInfo getDataInfo(String storedText) {
     if (TextUtils.isEmpty(storedText)) {
       throw new NullPointerException("Text should not be null or empty");
     }
-    int index = storedText.indexOf(DELIMITER);
+
+    int index = storedText.indexOf(INFO_DELIMITER, 0);
     if (index == -1) {
-      throw new IllegalArgumentException("Text should contain delimiter");
-    }
-    String text = storedText.substring(0, index);
-    String cipherText = storedText.substring(index + 1);
-    if (TextUtils.isEmpty(text) || TextUtils.isEmpty(cipherText)) {
-      throw new IllegalArgumentException("Invalid stored text");
-    }
-    char firstChar = text.charAt(text.length() - 1);
-    if (firstChar == NEW_VERSION) {
-      return getNewDataInfo(text, cipherText);
-    } else {
-      return getOldDataInfo(text, cipherText);
+      return getOldDataInfo(storedText);
     }
+    return getNewDataInfo(storedText, index);
   }
 
-  static DataInfo getNewDataInfo(String text, String cipherText) {
-    //first char defines whether it is new version or not
+  static DataInfo getNewDataInfo(String storedText, int indexFirstDelimiter) {
+    String keyClass = storedText.substring(0, indexFirstDelimiter);
+
+    int start = indexFirstDelimiter + 1;
+    int index = storedText.indexOf(INFO_DELIMITER, start);
+
+    if (index == -1) {
+      throw new IllegalArgumentException("Text should contain delimiter");
+    }
+
+    String valueClass = storedText.substring(start, index);
+
+    char dataType = storedText.charAt(++index);
+
+    //skip NEW_VERSION + DELIMITER
+    index += 3;
 
-    String[] infos = text.split(INFO_DELIMITER);
```

Comment:
a split version uses a regex to find matches. it visits the String till the end. removing if would save few cpu cycles.